src/models/CommonModel.ts

```diff
@@ -20,6 +20,102 @@ export class CommonModel extends CommonSchema<CommonModel> {
   extend?: string[];
   originalSchema?: Schema | boolean;
   
+  /**
+   * Retrieves data from originalSchema by given key
+   * 
+   * @param key given key
+   * @returns {any}
+   */
+  getFromSchema<K extends keyof Schema>(key: K) {
+    let schema = this.originalSchema || {};
+    if (typeof schema === 'boolean') schema = {};
+    return schema[`${key}`];
+  }
+
+  /**
+   * Set the types of the model
+   * 
+   * @param types to set the model type to
+   */
+  setType(types : string | string[]) {
+    if (Array.isArray(types)) {
+      if (types.length === 0) {
+        this.type = undefined;
+        return;
+      } else if (types.length === 1) {
+        this.type = types[0];
+        return;
+      }
+    }
+    this.type = types;
+  }
+
+  /**
+   * Adds types to the existing model types.
+   * 
+   * Makes sure to only keep a single type incase of duplicates.
+   * 
+   * @param types which types we should try and add to the existing output
+   */
+  addToTypes(types: string[] | string) {
```

Comment:
I do not like this `|` in arguments, it is not following Single Responsibility Principle.

Reply:
I don't see how this violates SRP. 

The function's single responsibility is to add type(s) to the model, we just allow one or more types to be added.

Why do you say it violates the SRP? 🤔

Reply:
Because you have multiple ways to execute this function and both of them give you same result. Also you need to add unit test to cover both cases. Few weeks ago colleague of mine ask me about exactly same case.

Now you need to write 2 tests and add a logic to handle both cases. But what I prefer to do is to pass even single element in the array.
So instead of calling it this way
```
addToTypes(SOME_TYPE);
```
you may want to do
```
addToTypes([SOME_TYPE]);
```

Reply:
SRP means that a function should be responsible only for smallest possible bit of whole process. This function above is responsible for 2 of them 
 - adding single element
 - adding array of elements
 
You may see in some libraries that this is something allowed. Yeah it is but those libraries are written and C++ (or violate SRP) where you can do arguments overload based on types of arguments.

Reply:
Yea I see your point, also in terms of https://github.com/asyncapi/generator-model-sdk/pull/184#discussion_r632272973. 

To me, it comes down to whether or not you want cleaner syntax i.e. `addToTypes(SOME_TYPE);` over `addToTypes([SOME_TYPE]);` by adding a single test to support it (since we use recursion). Cause In terms of coverage that is all it takes. 

Especially since TS does not allow overloading functions, if this was Java I would have split it into two functions to allow both parameters.

So this comes down to preference IMO, lets see what others think.

cc. @magicmatatjahu

Reply:
@jonaslagoni 
Well it is not only preferences, you will produce more codebase which you will have maintain. 
Let me try to make it more clear with more code examples.
```ts
const ensureArray = (arrayOrOne: any[] | any) => {
   if(arrayOfStringsOrSingleString instanceof Array) {
      return arrayOfStringsOrSingleString;
   }
   return [arrayOfStringsOrSingleString];
};
```
and in tests
```ts
it("test for one type", () => {
   const payload = {};
   
   const result = ensureArray(payload);
   
   expect(result).toBeEqual([payload])
});
it("test for multiple types", () => {
   const payload = [{}, {}];
   
   const result = ensureArray(payload);
   
   expect(result).toBeEqual(payload);
});
```


```ts
addToTypes(ensureArray("payload1"));
// or
addToTypes(esnureArray(["payload#1","payload#2]));
```
But you can easly get rid of ensureArray function and tests if you just do.
```ts
addToTypes(["payload1"]);
// or
addToTypes(["payload#1","payload#2]);
```
Which is way cleaner and more maintainable way because you do not produce code you need to test and making clear that `addToTypes` function has only one way of working.
Reducing those `if`s inside `addToTypes` function will also reduce complexity in bot statistic (I am pretty sure I saw some bot with complexity reports here).

Reply:
In TS you can write overloading functions but at the end you must write only one function like currently.

```ts
addToTypes(types: string);
addToTypes(types: string[]);
addToTypes(types: string[] | string) {
 // ...logic
}
``` 

I can agree with @czlowiek488 in some points, but this function with this shape is fine for me. We can reduce complexity moving normal string to array of strings

```ts
addToTypes(types: string[] | string) {
  types = Array.isArray(types) ? types : [types];
  types = types.filter(Boolean);
  let result = types;
  if (types.length === 0) result = undefined;
  if (types.length === 1) result = types[0];
  return result;
}
```

Reply:
Also I will never agree to let it be like here. From my point of view you violate SRP.

Reply:
I guess it violate even YAGNI rule because you have multiple ways to achieve the same effect.

Reply:
@czlowiek488 no one disputes that in other languages you would have split these functions up and overloaded them to provide a more maintainable code. 

However, with the nature of what the function does, how it is used, the underlying meaning of the `type` property, together with the limitations of TS this implementation looks fine for me as well. Especially since switching to only allowing array would make it more complex to understand IMO, and it would a discussion about potato/potato (given the nature of TS).

Always appreciate the feedback and feel free to prove me wrong with a PR/issue which clearly shows that maintainability increases and reduces complexity 🙂